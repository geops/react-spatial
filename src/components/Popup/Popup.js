import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { MdClose } from 'react-icons/md';
import OLMap from 'ol/Map';
import Feature from 'ol/Feature';
import Point from 'ol/geom/Point';
import { getCenter } from 'ol/extent';
import { unByKey } from 'ol/Observable';
import Button from '../Button';

const propTypes = {
  children: PropTypes.node.isRequired,
  map: PropTypes.instanceOf(OLMap).isRequired,
  feature: PropTypes.instanceOf(Feature),
  className: PropTypes.string,
  classNameCloseBt: PropTypes.string,
  onCloseClick: PropTypes.func,
  onKeyUp: PropTypes.func,
  showCloseButton: PropTypes.bool,

  t: PropTypes.func,
};

const defaultProps = {
  feature: null,
  className: 'tm-popup',
  classNameCloseBt: 'tm-button tm-popup-close-bt',
  showCloseButton: true,
  onKeyUp: () => {},
  onCloseClick: () => {},
  t: p => p,
};

class Popup extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      top: 0,
      left: 0,
    };
    this.postrenderKey = null;
  }

  componentDidMount() {
    const { map } = this.props;
    this.postrenderKey = map.on('postrender', () => {
      this.updatePixelPosition();
    });
  }

  componentDidUpdate() {
    const { feature } = this.props;
    if (feature) {
      // Initialize the position.
      this.updatePixelPosition();
    }
  }

  componentWillUnmount() {
    unByKey(this.postrenderKey);
  }

  updatePixelPosition() {
    const { map, feature } = this.props;
    if (feature) {
      let coord;
      const geom = feature.getGeometry();
      if (geom instanceof Point) {
        coord = geom.getCoordinates();
      } else {
        coord = getCenter(geom.getExtent());
      }
      const pos = map.getPixelFromCoordinate(coord);
      this.setState({
        left: pos[0],
        top: pos[1],
      });
    }
  }

  renderCloseButton() {
    const { t, showCloseButton, onCloseClick, classNameCloseBt } = this.props;

    if (!showCloseButton) {
      return null;
    }

    return (
      <Button
        className={classNameCloseBt}
        title={`Popup ${t('Schliessen')}`}
        onClick={() => onCloseClick()}
      >
        <MdClose focusable={false} />
      </Button>
    );
  }

  render() {
    const { feature, className, children, onKeyUp } = this.props;

    if (!feature) {
      return null;
    }

    const { top, left } = this.state;

    return (
      <div
        className={className}
        style={{
          position: 'absolute',
          left,
          top,
        }}
      >
        <div
          role="button"
          tabIndex={className === 'tm-tooltip' ? '0' : ''}
          onKeyUp={e => {
            onKeyUp(e);
          }}
        >
          {this.renderCloseButton()}
          {children}
        </div>
      </div>
    );
  }
}

Popup.propTypes = propTypes;
Popup.defaultProps = defaultProps;

export default Popup;
