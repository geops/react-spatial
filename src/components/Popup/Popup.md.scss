.tm-popup-example {
  position: relative;
  overflow: hidden;

  .tm-map {
    position: relative;
    height: 300px;
    width: 100%;
  }
}
