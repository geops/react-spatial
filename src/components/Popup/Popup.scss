.tm-popup > div {
  position: absolute;
  background-color: white;
  filter: drop-shadow(0 1px 4px rgba(0, 0, 0, 0.2));
  padding: 15px;
  bottom: 12px;
  left: -50px;
  min-width: 220px;

  &::after,
  &::before {
    top: 100%;
    border: solid transparent;
    content: ' ';
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &::after {
    border-top-color: white;
    border-width: 10px;
    left: 48px;
    margin-left: -10px;
  }

  &::before {
    border-top-color: #ccc;
    border-width: 11px;
    left: 48px;
    margin-left: -11px;
  }

  .tm-popup-close-bt {
    position: absolute;
    border: none;
    right: 0;
    top: 0;
    color: black;
    height: $btn-size-base;
    width: $btn-size-base;
  }

  .tm-popup-title {
    font-weight: bold;
    padding-bottom: 5px;
    margin-right: 25px;
  }
}
