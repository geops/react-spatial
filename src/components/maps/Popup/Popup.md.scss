.rs-popup-example {
  position: relative;
  overflow: hidden;
}

.rs-w-xs .rs-popup {
  left: 0 !important;
  right: 0;
  bottom: 0;
  top: auto !important;
  border: 1px solid #e8e8e8;
}

.rs-w-xs .rs-popup > div {
  position: relative;
  left: 0;
  bottom: 0;
  min-width: 0;

  &::after,
  &::before {
    display: none;
  }
}
