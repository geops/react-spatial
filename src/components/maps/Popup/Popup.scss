.rs-popup {
  position: absolute;

  .rs-popup-container {
    position: absolute;
    background-color: white;
    filter: drop-shadow($box-shadow);
    bottom: 12px;
    left: -50px;
    min-width: 220px;

    &::after,
    &::before {
      top: 100%;
      border: solid transparent;
      content: ' ';
      height: 0;
      width: 0;
      position: absolute;
      pointer-events: none;
    }

    &::after {
      border-top-color: white;
      border-width: 10px;
      left: 48px;
      margin-left: -10px;
    }

    &::before {
      border-top-color: #ccc;
      border-width: 11px;
      left: 48px;
      margin-left: -11px;
    }

    .rs-popup-close-bt {
      display: flex;
      align-items: center;
      justify-content: center;
    }

    .rs-popup-header {
      display: flex;
      align-items: center;
      justify-content: space-between;
      background-color: #f5f5f5;
      font-weight: bold;
      padding: 10px;
    }

    .rs-popup-body {
      padding: 10px;
    }
  }
}
