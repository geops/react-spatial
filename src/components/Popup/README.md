#

This demonstrates the use of Popup.

```jsx
import React from  'react'
import BasicMap from 'react-spatial/components/BasicMap';
import Popup from 'react-spatial/components/Popup';
import Layer from 'react-spatial/Layer';
import VectorLayer from 'react-spatial/VectorLayer';
import OLMap from 'ol/Map';
import Feature from 'ol/Feature';
import Point from 'ol/geom/Point';
import VectorSource from 'ol/source/Vector';
import TileLayer from 'ol/layer/Tile';
import TileGrid from 'ol/tilegrid/TileGrid';
import TileImageSource from 'ol/source/TileImage';
import { getCenter } from 'ol/extent';
import Style from 'ol/style/Style';
import Circle from 'ol/style/Circle';
import Fill from 'ol/style/Fill';

class PopupExample extends React.Component {
  constructor(props) {
    super(props);

    const extent = [599500, 199309, 600714, 200002];
    const resolutions = [
      6.927661,
      3.4638305,
      1.73191525,
      0.865957625,
      0.4329788125,
      0.21648940625,
      0.108244703125,
    ];

    const layer = new Layer({
      name: 'Layer',
      olLayer: new TileLayer({
        extent: extent,
        source: new TileImageSource({
          tileUrlFunction: c =>
            '//plans.trafimage.ch/static/tiles/' +
            `bern_aussenplan/${c[0]}/${c[1]}/${c[2]}.png`,
          tileGrid: new TileGrid({
            origin: [extent[0], extent[1]],
            resolutions: resolutions,
          }),
        }),
      }),
    });
    const vectorLayer = new VectorLayer({
      source: new VectorSource({
        features: [
          new Feature({
            geometry: new Point(getCenter(extent)),
          }),
        ],
      }),
      style: new Style({
        image: new Circle({
          radius: 40,
          fill: new Fill({
            color: '#ff0000',
          }),
        }),
      }),
    });
    this.center = getCenter(extent);
    this.layers = [layer, vectorLayer];
    this.map = new OLMap();
    this.onFeaturesClick = this.onFeaturesClick.bind(this);
    this.onCloseClick = this.onCloseClick.bind(this);

    this.state = {
      featureClicked: null,
    };
  }

  onFeaturesClick(features) {
    this.setState({
      featureClicked: features.length ? features[0] : null,
    });
  }

  onCloseClick() {
    this.setState({ featureClicked: null });
  }

  render() {
    const { featureClicked } = this.state;
    const content =
      featureClicked &&
      featureClicked
        .getGeometry()
        .getCoordinates()
        .toString();

    return (
      <div className="tm-popup-example">
        <BasicMap
          map={this.map}
          center={this.center}
          zoom={17}
          layers={this.layers}
          onFeaturesClick={this.onFeaturesClick}
        />
        <Popup
          map={this.map}
          feature={featureClicked}
          onCloseClick={this.onCloseClick}
        >
          <div>Geometry:</div>
          <div>{content}</div>
        </Popup>
      </div>
    );
  }
}

<PopupExample />;
```
