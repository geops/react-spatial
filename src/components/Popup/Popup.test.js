import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-canvas-mock';

import OLMap from 'ol/Map';
import Feature from 'ol/Feature';
import Point from 'ol/geom/Point';
import Line from 'ol/geom/LineString';
import Popup from './Popup';

const map = new OLMap({});

configure({ adapter: new Adapter() });

const feat = new Feature({
  geometry: new Point(0, 0),
});

const featLine = new Feature({
  geometry: new Line([[0, 0], [1, 1]]),
});

describe('Popup', () => {
  describe('should match snapshot', () => {
    test('without feature', () => {
      const component = renderer.create(
        <Popup map={map}>
          <div id="foo" />
        </Popup>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('with default values.', () => {
      const component = renderer.create(
        <Popup map={map} feature={feat}>
          <div id="foo" />
        </Popup>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('without close button.', () => {
      const component = renderer.create(
        <Popup map={map} feature={feat} showCloseButton={false}>
          <div id="bar" />
        </Popup>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('with tm-tooltip class.', () => {
      const component = renderer.create(
        <Popup map={map} feature={feat} className="tm-tooltip">
          <div id="bar" />
        </Popup>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  [['click', {}], ['keypress', { which: 13 }]].forEach(evt => {
    test(`should trigger onCloseClick function on ${evt[0]} event.`, () => {
      class Closer {
        static onCloseClick() {}
      }

      const spy = jest.spyOn(Closer, 'onCloseClick');

      const component = mount(
        <Popup
          map={map}
          feature={feat}
          onCloseClick={() => Closer.onCloseClick()}
          showCloseButton
        >
          <div id="gux" />
        </Popup>,
      );

      component
        .find('.tm-popup-close-bt')
        .at(1)
        .simulate(...evt);

      expect(spy).toHaveBeenCalled();
    });

    test(`should trigger default onCloseClick function on ${
      evt[0]
    } event without errors.`, () => {
      const component = mount(
        <Popup map={map} feature={feat} n>
          <div id="gux" />
        </Popup>,
      );
      // test if no js error triggered by the default value
      try {
        component
          .find('.tm-popup-close-bt')
          .at(1)
          .simulate(...evt);
        expect(true).toBe(true);
      } catch (e) {
        expect(false).toBe(true);
      }
    });
  });

  describe(`updates position`, () => {
    [feat, featLine].forEach(f => {
      test(`on feature's update.`, () => {
        map.getPixelFromCoordinate = jest.fn(() => [10, 200]);
        const component = shallow(
          <Popup map={map}>
            <div id="gux" />
          </Popup>,
        );
        const spy = jest.spyOn(component.instance(), 'updatePixelPosition');

        component.setProps({
          feature: f,
        });

        expect(spy).toHaveBeenCalledTimes(2);
      });
    });

    test(`on map postrender event.`, () => {
      map.getPixelFromCoordinate = jest.fn(() => [10, 200]);
      const component = shallow(
        <Popup map={map} feature={feat}>
          <div id="gux" />
        </Popup>,
      );
      const spy = jest.spyOn(component.instance(), 'updatePixelPosition');
      map.dispatchEvent({ type: 'postrender' });
      expect(spy).toHaveBeenCalledTimes(2);
    });
  });

  test(`deregisters postrender on unmount.`, () => {
    map.getPixelFromCoordinate = jest.fn(() => [10, 200]);
    const component = shallow(
      <Popup map={map} feature={feat}>
        <div id="gux" />
      </Popup>,
    );
    const spy = jest.spyOn(component.instance(), 'updatePixelPosition');
    map.dispatchEvent({ type: 'postrender' });
    expect(spy).toHaveBeenCalledTimes(2);
    component.unmount();
    map.dispatchEvent({ type: 'postrender' });
    expect(spy).toHaveBeenCalledTimes(2);
  });
});
