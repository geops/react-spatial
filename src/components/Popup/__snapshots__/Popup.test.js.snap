// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup should match snapshot with default values. 1`] = `
<div
  className="tm-popup"
  style={
    Object {
      "left": 0,
      "position": "absolute",
      "top": 0,
    }
  }
>
  <div
    onKeyUp={[Function]}
    role="button"
    tabIndex=""
  >
    <div
      aria-label="Popup Schliessen"
      className="tm-button tm-popup-close-bt"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title="Popup Schliessen"
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      id="foo"
    />
  </div>
</div>
`;

exports[`Popup should match snapshot with tm-tooltip class. 1`] = `
<div
  className="tm-tooltip"
  style={
    Object {
      "left": 0,
      "position": "absolute",
      "top": 0,
    }
  }
>
  <div
    onKeyUp={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      aria-label="Popup Schliessen"
      className="tm-button tm-popup-close-bt"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      title="Popup Schliessen"
    >
      <svg
        fill="currentColor"
        focusable={false}
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
    <div
      id="bar"
    />
  </div>
</div>
`;

exports[`Popup should match snapshot without close button. 1`] = `
<div
  className="tm-popup"
  style={
    Object {
      "left": 0,
      "position": "absolute",
      "top": 0,
    }
  }
>
  <div
    onKeyUp={[Function]}
    role="button"
    tabIndex=""
  >
    <div
      id="bar"
    />
  </div>
</div>
`;

exports[`Popup should match snapshot without feature 1`] = `null`;
